### PR TITLE
Custom db driver.

### PR DESCRIPTION
**Pull request by makasim**
Right now, the bundle is coupled to some drivers, there is no easy way use it with something else.

This PR makes it possible to do next, the `service` section become mandatory if custom db driver is selected. 

```yaml
fos_oauth_server:
  db_driver: custom
  client_class: App\Infra\OAuth\Model\OAuthClient
  access_token_class: App\Infra\OAuth\Model\OAuthAccessToken
  refresh_token_class: App\Infra\OAuth\Model\OAuthRefreshToken
  auth_code_class: App\Infra\OAuth\Model\OAuthAuthCode
  service:
    user_provider: App\Infra\OAuth\YadmUserProvider
    client_manager: App\Infra\OAuth\YadmClientManager
    access_token_manager: App\Infra\OAuth\YadmAccessTokenManager
    refresh_token_manager: App\Infra\OAuth\YadmRefreshTokenManager
    auth_code_manager: App\Infra\OAuth\YadmAuthCodeManager
    options:
        access_token_lifetime: 604800 # 7days
```

### CODE REVIEW

**Comment by makasim** [comment]
@dkarlovi wdyt?

File: DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -30,9 +30,38 @@ public function getConfigTreeBuilder()
         $treeBuilder = new TreeBuilder();
         $rootNode = $treeBuilder->root('fos_oauth_server');
 
-        $supportedDrivers = array('orm', 'mongodb', 'propel');
+        $supportedDrivers = array('orm', 'mongodb', 'propel', 'custom');
```
Comment (dkarlovi) [inline DependencyInjection/Configuration.php]: Could we have some docs about the feature?

File: Resources/config/custom.xml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+
+<container xmlns="http://symfony.com/schema/dic/services"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+    <services>
+        <!-- you should define all services from fos_oauth_server.service configuration section explicitly -->
```
Comment (dkarlovi) [inline Resources/config/custom.xml]: Since we're adding this empy file, maybe add a (commented-out) sample configuration of required services? We could at least enumerate them here.

File: Resources/config/oauth.xml [hunk 3]
```diff
@@ -37,7 +37,7 @@
         </service>
 
         <service id="fos_oauth_server.create_client_command" class="FOS\OAuthServerBundle\Command\CreateClientCommand">
-            <argument type="service" id="fos_oauth_server.client_manager.default" />
+            <argument type="service" id="fos_oauth_server.client_manager" />
```
Comment (dkarlovi) [inline Resources/config/oauth.xml]: Why was this required?

Comment (makasim) [reply Resources/config/oauth.xml]: It is driver implementation detail. The bundle should use services without `.default` postfix. It follows the rule in other places for other services. This is the only place where the rule is broken.

File: Resources/config/custom.xml (hunk 2 above)
Comment (stof) [reply Resources/config/custom.xml]: don't create this file at all (and handle this when loading the file in the DI extension)

File: DependencyInjection/Configuration.php (hunk 1 above)
Comment (dkarlovi) [reply DependencyInjection/Configuration.php]: The description of the PR is a good start IMO.

File: Resources/config/custom.xml (hunk 2 above)
Comment (makasim) [reply Resources/config/custom.xml]: We would have to add if condition to cover the special case, I am in favor of keeping the file and adding the example

Comment (makasim) [reply Resources/config/custom.xml]: What should I do?

Comment (makasim) [reply Resources/config/custom.xml]: Ok, I'll do

**Comment by makasim** [comment]
@dkarlovi please review.

File: Resources/doc/custom_db_driver.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
+Custom db driver.
+=================
+
+The bundle provides drivers for Doctrine ORM, Doctrine MongoDB, and Propel libraries.
+Though sometimes you might want to use the bundle with a custom or in house written storage.
```
Comment (dkarlovi) [inline Resources/doc/custom_db_driver.md]: > in-house

File: Resources/doc/custom_db_driver.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+For that the bundle has a custom storage.
```
Comment (dkarlovi) [inline Resources/doc/custom_db_driver.md]: > For that, the bundle has support for custom storage.

File: Resources/doc/custom_db_driver.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
+Custom db driver.
+=================
+
+The bundle provides drivers for Doctrine ORM, Doctrine MongoDB, and Propel libraries.
+Though sometimes you might want to use the bundle with a custom or in house written storage.
+For that the bundle has a custom storage. 
+Once set it makes manager options in fos_oauth_server.service section mandatory.
```
Comment (dkarlovi) [inline Resources/doc/custom_db_driver.md]: > Once set, setting manager options in `fos_oauth_server.service` section becomes mandatory.

File: Resources/doc/custom_db_driver.md [hunk 7]
```diff
@@ -0,0 +1,27 @@
+Custom db driver.
+=================
+
+The bundle provides drivers for Doctrine ORM, Doctrine MongoDB, and Propel libraries.
+Though sometimes you might want to use the bundle with a custom or in house written storage.
+For that the bundle has a custom storage. 
+Once set it makes manager options in fos_oauth_server.service section mandatory.  
+They have to be set.
```
Comment (dkarlovi) [inline Resources/doc/custom_db_driver.md]: Drop this.

File: Tests/DependencyInjection/FOSOAuthServerExtensionTest.php [hunk 8]
```diff
@@ -9,9 +9,11 @@
  * file that was distributed with this source code.
  */
 
-namespace DependencyInjection;
+namespace FOS\OAuthServerBundle\Tests\DependencyInjection;
```
Comment (dkarlovi) [inline Tests/DependencyInjection/FOSOAuthServerExtensionTest.php]: Nice. :+1:

**Comment by makasim** [comment]
@dkarlovi updated

**Comment by dkarlovi** [comment]
@makasim one test failed again, it's a case mismatch.

**Comment by makasim** [comment]
@dkarlovi green

**Comment by dkarlovi** [comment]
LGTM.

**Comment by dkarlovi** [comment]
@makasim thanks!
